lib/media/track/mediatrack.js

```diff
@@ -237,4 +271,82 @@ class MediaTrack extends Track {
   }
 }
 
+/**
+ * Play an HTMLMediaElement when it is paused and not backgrounded.
+ * @private
+ * @param {HTMLMediaElement} el
+ * @param {Log} log
+ * @returns {void}
+ */
+function playWhenPausedAndNotBackgrounded(el, log) {
+  const tag = el.tagName.toLowerCase();
+  log.warn('Unintentionally paused:', el);
+
+  // NOTE(mmalavalli): When the element is unintentionally paused, we wait one
+  // second for the "onvisibilitychange" event on the HTMLDocument to see if the
+  // app will be backgrounded. If not, then the element can be safely played.
+  Promise.race([
+    waitForEvent(document, 'visibilitychange'),
+    waitForSometime(1000)
+  ]).then(() => {
+    if (document.visibilityState === 'visible') {
+      log.info(`Playing unintentionally paused <${tag}> element`);
+      log.debug('Element:', el);
+      el.play().then(() => {
+        log.info(`Successfully played unintentionally paused <${tag}> element`);
+        log.debug('Element:', el);
+      }).catch(error => {
+        log.warn(`Error while playing unintentionally paused <${tag}> element:`, error, el);
+      });
+    }
+  });
+}
+
+/**
+ * Shim the pause() and play() methods of the given HTMLMediaElement so that
+ * we can detect if it was paused unintentionally.
+ * @param {HTMLMediaElement} el
+ * @param {?function} [onUnintentionallyPaused=null]
+ * @returns {{pausedIntentionally: function, unShim: function}}
+ */
+function shimMediaElement(el, onUnintentionallyPaused = null) {
+  const origPause = el.pause;
+  const origPlay = el.play;
+
+  let pausedIntentionally = false;
+
+  el.pause = () => {
+    pausedIntentionally = true;
+    return origPause.call(el);
+  };
+
+  el.play = () => {
+    pausedIntentionally = false;
+    return origPlay.call(el);
+  };
+
+  const onPause = onUnintentionallyPaused ? () => {
+    if (!pausedIntentionally) {
+      onUnintentionallyPaused();
+    }
+  } : null;
+
+  if (onPause) {
+    el.addEventListener('pause', onPause);
+  }
+
+  return {
+    pausedIntentionally() {
+      return pausedIntentionally;
+    },
+    unShim() {
+      el.pause = origPause;
+      el.play = origPlay;
+      if (onPause) {
+        el.removeEventListener('pause', onPause);
+      }
+    }
+  };
+}
```

Comment:
This would look simpler if you change the default value for `onUnintentionallyPaused`

```suggestion
/**
 * Shim the pause() and play() methods of the given HTMLMediaElement so that
 * we can detect if it was paused unintentionally.
 * @param {HTMLMediaElement} el
 * @param {?function} [onUnintentionallyPaused=() => {}]
 * @returns {{pausedIntentionally: function, unShim: function}}
 */
function shimMediaElement(el, onUnintentionallyPaused = () => {}) {
  const origPause = el.pause;
  const origPlay = el.play;

  let pausedIntentionally = false;

  el.pause = () => {
    pausedIntentionally = true;
    return origPause.call(el);
  };

  el.play = () => {
    pausedIntentionally = false;
    return origPlay.call(el);
  };

  const onPause = () => {
    if (!pausedIntentionally) {
      onUnintentionallyPaused();
    }
  };

  el.addEventListener('pause', onPause);

  return {
    pausedIntentionally() {
      return pausedIntentionally;
    },
    unShim() {
      el.pause = origPause;
      el.play = origPlay;
        el.removeEventListener('pause', onPause);
    }
  };
}
```

Reply:
@makarandp0 Although this reads better, I don't want to add unnecessary pause event handlers.